source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc

```diff
@@ -0,0 +1,184 @@
+#include "extensions/filters/http/header_to_metadata/header_to_metadata_filter.h"
+
+#include "common/config/well_known_names.h"
+#include "common/protobuf/protobuf.h"
+
+#include "extensions/filters/http/well_known_names.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace HeaderToMetadataFilter {
+namespace {
+
+const uint32_t MAX_HEADER_VALUE_LEN = 100;
+
+} // namespace
+
+Config::Config(const envoy::config::filter::http::header_to_metadata::v2::Config config) {
+  request_set_ = Config::configToVector(config.request_rules(), request_rules_);
+  response_set_ = Config::configToVector(config.response_rules(), response_rules_);
+
+  // don't allow an empty configuration
+  if (!response_set_ && !request_set_) {
+    throw new EnvoyException("Must at least specify either response or request config");
+  }
+}
+
+bool Config::configToVector(const ProtobufRepeatedRule& protoRules, HeaderToMetadataRules& vector) {
+  if (protoRules.size() == 0) {
+    ENVOY_LOG(debug, "no rules provided");
+    return false;
+  }
+
+  for (const auto& entry : protoRules) {
+    std::pair<Http::LowerCaseString, Rule> rule = {Http::LowerCaseString(entry.header()), entry};
+
+    // Rule must have at least one of the `on_header_*` fields set.
+    if (!entry.has_on_header_present() && !entry.has_on_header_missing()) {
+      const auto& error = fmt::format("header to metadata filter: rule for header '{}' has neither "
+                                      "`on_header_present` nor `on_header_missing` set",
+                                      entry.header());
+      throw EnvoyException(error);
+    }
+
+    vector.push_back(rule);
+  }
+
+  return true;
+}
+
+HeaderToMetadataFilter::HeaderToMetadataFilter(const ConfigSharedPtr config) : config_(config) {}
+
+HeaderToMetadataFilter::~HeaderToMetadataFilter() {}
+
+Http::FilterHeadersStatus HeaderToMetadataFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  if (config_->doRequest()) {
+    writeHeaderToMetadata(headers, config_->requestRules(), *decoder_callbacks_);
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+void HeaderToMetadataFilter::setDecoderFilterCallbacks(
+    Http::StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+}
+
+Http::FilterHeadersStatus HeaderToMetadataFilter::encodeHeaders(Http::HeaderMap& headers, bool) {
+  if (config_->doResponse()) {
+    writeHeaderToMetadata(headers, config_->responseRules(), *encoder_callbacks_);
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+void HeaderToMetadataFilter::setEncoderFilterCallbacks(
+    Http::StreamEncoderFilterCallbacks& callbacks) {
+  encoder_callbacks_ = &callbacks;
+}
+
+bool HeaderToMetadataFilter::addMetadata(StructMap& map, const std::string& metaNamespace,
```

Comment:
Sorry I missed this before, but our variable style is all-lowercase/snake_case, so "meta_namespace".